File: code/java/interview_arena_article-04.java
public class Interview_arena_article_04 {
    /**
     * Binary search efficiently determines subsequences
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 2) {
            throw new IllegalArgumentException("Expected: (String s, String t)");
        }

        Solution solution = new Solution();
        return solution.isSubsequence((String) args[0], (String) args[1]);
    }

    private static class Solution {

        public boolean isSubsequence(String s, String t) {
            int i = 0, j = 0;
            while (i < s.length() && j < t.length()) {
                if (s.charAt(i) == t.charAt(j)) {
                    i++;
                }
                j++;
            }
            return i == s.length();
        }
    }
}

File: code/python/interview_arena_article-04.py
"""Binary search efficiently determines subsequences"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().isSubsequence(*args, **kwargs)


class _Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        i, j = 0, 0
        while i < len(s) and j < len(t):
            if s[i] == t[j]:
                i += 1
            j += 1
        return i == len(s)
